nexus/src/db/datastore.rs

```diff
@@ -238,6 +242,173 @@ impl DataStore {
         })
     }
 
+    /// Gets allocated regions for a disk, and the datasets to which those
+    /// regions belong.
+    ///
+    /// Note that this function does not validate liveness of the Disk, so it
+    /// may be used in a context where the disk is being deleted.
+    pub async fn get_allocated_regions(
+        &self,
+        disk_id: Uuid,
+    ) -> Result<Vec<(Dataset, Region)>, Error> {
+        use db::schema::dataset::dsl as dataset_dsl;
+        use db::schema::region::dsl as region_dsl;
+        region_dsl::region
+            .filter(region_dsl::disk_id.eq(disk_id))
+            .inner_join(
+                dataset_dsl::dataset
+                    .on(region_dsl::dataset_id.eq(dataset_dsl::id)),
+            )
+            .select((Dataset::as_select(), Region::as_select()))
+            .get_results_async::<(Dataset, Region)>(self.pool())
+            .await
+            .map_err(|e| public_error_from_diesel_pool_shouldnt_fail(e))
+    }
+
+    /// Idempotently allocates enough regions to back a disk.
+    ///
+    /// Returns the allocated regions, as well as the datasets to which they
+    /// belong.
+    pub async fn region_allocate(
+        &self,
+        disk_id: Uuid,
+        params: &params::DiskCreate,
+    ) -> Result<Vec<(Dataset, Region)>, Error> {
+        use db::schema::dataset::dsl as dataset_dsl;
+        use db::schema::region::dsl as region_dsl;
+
+        // ALLOCATION POLICY
```

Comment:
I struggled a little to find a balance with this. I wanted something "real-ish", to show how we *can* access a lot of data, but not *too* complicated, so that it's easy to change.